lib/extensions/outline.py

```diff
@@ -3,48 +3,79 @@
+        self.arg_parser.add_argument("-t", "--threshold", type=float, default=10.0, dest="threshold")
+        self.arg_parser.add_argument("-i", "--inset", type=float, default=0.001, dest="inset")
 
     def effect(self):
         if not self.svg.selection:
-            inkex.errormsg(_("Please select one or more shapes to convert to their outline."))
+            errormsg(_("Please select one or more shapes to convert to their outline."))
             return
 
+        self.threshold = self.options.threshold * PIXELS_PER_MM
+        self.shape_buffer = max(self.options.buffer * PIXELS_PER_MM, 0.001)
+        self.smoothness = self.options.smoothness * PIXELS_PER_MM
+        self.inset = self.options.inset * PIXELS_PER_MM
+
         for element in self.svg.selection:
             self.element_to_outline(element)
 
+    def get_outline(self, element):
+        d = ''
+        transform = element.composed_transform()
+        path = Path(element.get_path()).transform(transform).break_apart()
+        for subpath in path:
+            points = subpath.end_points
+            shape = LineString(points).buffer(self.shape_buffer)
+            outline = ensure_multi_line_string(offset_curve(shape, -self.inset))
+
+            interiors = []
+            for interior in outline.geoms:
+                if Polygon(interior).area < self.threshold:
+                    continue
+                interior_path = smooth_path(interior.coords, self.smoothness)
+                if len(interior_path) > 2:
+                    interiors.append(Polygon(interior_path))
+            outline = MultiPolygon(interiors)
+
+            for geom in outline.geoms:
+                d += str(Path(geom.exterior.coords).transform(-transform))
+        return d
+
     def element_to_outline(self, element):
         if element.tag_name == 'g':
             for element in element.iterdescendants(SVG_PATH_TAG):
                 self.element_to_outline(element)
             return
-
-        path = element.get_path()
-        path = path.end_points
-        hull = concave_hull(LineString(path), ratio=self.options.ratio, allow_holes=self.options.allow_holes)
-        if isinstance(hull, LineString):
+        elif element.tag_name != 'path':
+            element_id = element.label or element.get_id()
+            errormsg(_(f"{element_id} is not a path element. "
+                       "This extension is designed to generate an outline of an embroidery pattern."))
             return
 
-        if not isinstance(hull, MultiPolygon):
-            hull = MultiPolygon([hull])
-        d = ''
-        for geom in hull.geoms:
-            d += 'M '
-            for x, y in geom.exterior.coords:
-                d += f'{x}, {y} '
-            d += "Z"
+        d = self.get_outline(element)
+        if not d:
+            errormsg(_("Could not generate path from element with the given settings."))
```

Comment:
You probably want to `return` out here, as otherwise it puts empty paths in the document after showing this error message.

Reply:
Thank you for testing!!